src/system_call_hooking.c

```diff
@@ -0,0 +1,478 @@
+            ((real_flags & MS_RDONLY) && !(real_flags & MS_REMOUNT))) {
+                // bind, shared, move, or new read-only mounts it do not affect
+                // the state of the driver
+                sys_ret = orig_mount(dev_name, dir_name, type, flags, data);
+        } else if ((real_flags & MS_RDONLY) && (real_flags & MS_REMOUNT)) {
+                // we are remounting read-only, same as umounting as far as the
+                // driver is concerned
+                ret = handle_bdev_mount_nowrite(dir_name, 0, &idx);
+                sys_ret = orig_mount(dev_name, dir_name, type, flags, data);
+                post_umount_check(ret, sys_ret, idx, dir_name);
+        } else {
+                // new read-write mount
+                sys_ret = orig_mount(dev_name, dir_name, type, flags, data);
+                if (!sys_ret)
+                        handle_bdev_mounted_writable(dir_name, &idx);
+        }
+
+        LOG_DEBUG("mount returned: %ld", sys_ret);
+
+        return sys_ret;
+}
+
+asmlinkage long umount_hook(char __user *name, int flags)
+{
+        int ret;
+        long sys_ret;
+        unsigned int idx;
+        char *buff_dev_name = NULL;
+
+        buff_dev_name = kmalloc(PATH_MAX, GFP_ATOMIC);
+        if (!buff_dev_name) {
+                return -ENOMEM;
+        }
+        ret = copy_from_user(buff_dev_name, name, PATH_MAX);
+        if (ret)
+                LOG_DEBUG("detected block device umount error:%d", ret);
+        else
+                LOG_DEBUG("detected block device umount: %s : %d",
+                          buff_dev_name, flags);
+        kfree(buff_dev_name);
+
+        ret = handle_bdev_mount_nowrite(name, flags, &idx);
+        sys_ret = orig_umount(name, flags);
+        post_umount_check(ret, sys_ret, idx, name);
+
+        LOG_DEBUG("umount returned: %ld", sys_ret);
+
+        return sys_ret;
+}
+
+#ifdef HAVE_SYS_OLDUMOUNT
+asmlinkage long oldumount_hook(char __user *name)
+{
+        int ret;
+        long sys_ret;
+        unsigned int idx;
+        char *buff_dev_name = NULL;
+
+        buff_dev_name = kmalloc(PATH_MAX, GFP_ATOMIC);
+        if (!buff_dev_name) {
+                return -ENOMEM;
+        }
+        ret = copy_from_user(buff_dev_name, name, PATH_MAX);
+        if (ret)
+                LOG_DEBUG("detected block device oldumount error:%d", ret);
+        else
+                LOG_DEBUG("detected block device oldumount: %s", name);
+        kfree(buff_dev_name);
+
+        ret = handle_bdev_mount_nowrite(name, 0, &idx);
+        sys_ret = orig_oldumount(name);
+        post_umount_check(ret, sys_ret, idx, name);
+
+        LOG_DEBUG("oldumount returned: %ld", sys_ret);
+
+        return sys_ret;
+}
+#endif
+
+inline void disable_page_protection(unsigned long *cr0)
+{
+        *cr0 = read_cr0();
+        write_cr0(*cr0 & ~X86_CR0_WP);
+}
+
+inline void reenable_page_protection(unsigned long *cr0)
+{
+        write_cr0(*cr0);
+}
```

Comment:
These two and their dependent definitions will be used for the 5.9 fix outside of this module. Up to you whether you want to split them out now, or we can do the split later as part of the fix MR.